tests/test_advection.py

```diff
@@ -1,167 +1,224 @@
+    for name, err in errors.items():
+        if err > tol:
+            print("Test fails for field %s with options ibp = %s, continuity = %s" % (name, dict(field_eqns)[name].ibp, dict(field_eqns[name]).continuity))
+    assert(all([err < tol for err in errors.values()]))
 
-    f_err = Function(f.function_space()).assign(f_end - f)
-    return f_err
 
-
-@pytest.mark.parametrize("geometry", ["slice", "sphere"])
 @pytest.mark.parametrize("time_discretisation", ["ssprk", "implicit_midpoint"])
-@pytest.mark.parametrize("ibp", ["once", "twice"])
-@pytest.mark.parametrize("equation_form", ["advective", "continuity"])
 @pytest.mark.parametrize("vector", [False, True])
-def test_advection_dg(tmpdir, geometry, time_discretisation, ibp, equation_form, vector):
+def test_advection_2Dsphere(tmpdir, time_discretisation, vector):
 
+    # set up physical_domain and state
+    physical_domain = Sphere(radius=1., ref_level=2)
     dirname = str(tmpdir)
-    f_err = run(dirname, geometry, time_discretisation, ibp, equation_form, vector)
-    assert(abs(f_err.dat.data.max()) < error[geometry])
+    fieldlist = ["f1", "f2", "f3", "f4"]
+    output = OutputParameters(dirname=dirname, dumplist=fieldlist, dumpfreq=15)
+    state = AdvectionDiffusionState(physical_domain.mesh,
+                                    horizontal_degree=1, family="BDM",
+                                    output=output)
+
+    # initial conditions, depending on whether we're advecting a
+    # vector or scalar field
+    x = SpatialCoordinate(physical_domain.mesh)
+    if vector:
+        fspace = VectorFunctionSpace(physical_domain.mesh, "DG", 1)
+        fexpr = as_vector([exp(-x[2]**2 - x[1]**2), 0., 0.])
+    else:
+        fspace = state.spaces("DG")
+        fexpr = exp(-x[2]**2 - x[1]**2)
+    for fname in fieldlist:
+        f = state.fields(fname, fspace)
+        f.interpolate(fexpr)
+
+    # advecting velocity
+    uspace = state.spaces("HDiv")
+    u0 = state.fields("u", uspace)
+    u0.project(as_vector([-x[1], x[0], 0.0]))
+
+    # timestepping parameters
+    timestepping = TimesteppingParameters(dt=pi/3*0.01)
+
+    # equations - we are testing both forms of the advection equation
+    # (advective and continuity) and integrating by parts once and
+    # twice for each option - this gives us 4 combinations.
+    eqnlist = []
+    eqnlist.append(AdvectionEquation(physical_domain, fspace, uspace))
+    eqnlist.append(AdvectionEquation(physical_domain, fspace, uspace, ibp="twice"))
+    eqnlist.append(AdvectionEquation(physical_domain, fspace, uspace, equation_form="continuity"))
+    eqnlist.append(AdvectionEquation(physical_domain, fspace, uspace, ibp="twice", equation_form="continuity"))
+
+    # setup the model with the above equations and the time_discretisation
+    field_eqns = list(zip(fieldlist, eqnlist))
+    model = setup_model(state, physical_domain, timestepping, field_eqns, time_discretisation)
+
+    # run the model
+    f_dict = run(model, pi/2., fieldlist)
```

Comment:
So `Timestepper.run` returns a dict of the fields you asked for in `fieldlist` if its not `None`.  That seems really odd.  Why not just do:
```python
fields = [(name, gettattr(state.fields, field)) for name in fieldlist]
```
here.

I note in passing that this syntax could be made nice with a small change to `FieldCreator` to something like:

```
class FieldCreator(OrderedDict):
    def __init__(self, fieldlist=None, xn=None, dumplist=None, pickup=True):
        super().__init__()
        for name, func in zip(fieldlist, xn.split()):
            self.add_field(func, name, name in dumplist, pickup)

    def add_field(self, field, name, dump, pickup):
        self[name] = field
        field.dump = dump
        field.pickup = pickup
        field.rename(name)
        return field
        
    def __getattr__(self, name):
        try:
            return self[name]
        except KeyError as e:
            raise AttributeError("Unknown field %s" % name) from e
    
    def __call__(self, name, space=None, dump=True, pickup=True):
        try:
            return self[name]
        except KeyError:
            f = Function(space)
            return self.add_field(f, name, dump, pickup)
    
    def __iter__(self):
        return iter(self.values())
```

now you can either do `state.fields[name]` or `getattr(state.fields, name)` or `state.fields.foo` or `state.fields["foo"]`